src/caffe/layers/memory_data_layer.cpp

```diff
@@ -117,6 +119,19 @@ void MemoryDataLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
     has_new_data_ = false;
 }
 
+template <typename Dtype>
+void MemoryDataLayer<Dtype>::Forward_gpu(const vector<Blob<Dtype>*>& bottom,
+      const vector<Blob<Dtype>*>& top) {
+  CHECK(data_) << "MemoryDataLayer needs to be initalized by calling Reset";
+  top[0]->Reshape(batch_size_, channels_, height_, width_);
+  top[1]->Reshape(batch_size_, 1, 1, 1);
```

Comment:
I can't figure out why this Reshape code is in Forward... it looks like it was added in https://github.com/BVLC/caffe/commit/cedefd77625db978623fcf00ecc32d899eed954b, along with #1416, but I don't see why we don't just do the reshape at the time the batch size is changed? That would avoid duplicating these lines between CPU and GPU.